src/Storages/ObjectStorage/StorageObjectStorageStableTaskDistributor.cpp

```diff
@@ -0,0 +1,175 @@
+#include "StorageObjectStorageStableTaskDistributor.h"
+#include <Common/SipHash.h>
+#include <consistent_hashing.h>
+#include <optional>
+
+namespace DB
+{
+
+StorageObjectStorageStableTaskDistributor::StorageObjectStorageStableTaskDistributor(
+    std::shared_ptr<IObjectIterator> iterator_,
+    std::optional<std::vector<std::string>> ids_of_nodes_)
+    : iterator(std::move(iterator_))
+    , connection_to_files(ids_of_nodes_.has_value() ? ids_of_nodes_.value().size() : 1)
+    , ids_of_nodes(ids_of_nodes_)
+    , iterator_exhausted(false)
+{
+}
+
+std::optional<String> StorageObjectStorageStableTaskDistributor::getNextTask(size_t number_of_current_replica)
+{
+    LOG_TRACE(
+        log,
+        "Received a new connection from replica {} looking for a file",
+        number_of_current_replica
+    );
+
+    // 1. Check pre-queued files first
+    if (auto file = getPreQueuedFile(number_of_current_replica))
+        return file;
+
+    // 2. Try to find a matching file from the iterator
+    if (auto file = getMatchingFileFromIterator(number_of_current_replica))
+        return file;
+
+    // 3. Process unprocessed files if iterator is exhausted
+    return getAnyUnprocessedFile(number_of_current_replica);
+}
+
+size_t StorageObjectStorageStableTaskDistributor::getReplicaForFile(const String & file_path)
+{
+    if (!ids_of_nodes.has_value())
+        return 0;
+
+    /// Trivial case
+    if (ids_of_nodes.value().size() < 2)
```

Comment:
Nitpick: extract ids_of_nodes into a non-optonal to avoid de-referencing all the time

Reply:
This is optional in `IStorageCluster` because all except `StorageObjectStorageCluster` do not use it.
To make it non-optional need to make a copy of list inside class. Or use smart pointer, than need to check if pointer has value.
Remove some dereferencings.

Reply:
can't you just do the below inside this method?

`auto ids_of_nodes = ids_of_nodes_optional.value_or_die()`

and use `ids_of_nodes` instead of repeating `ids_of_nodes.value()`?

Reply:
Ah, you updated the PR with that here: https://github.com/Altinity/ClickHouse/pull/709/commits/46fe767c2b08a982fed5fde4ee7c34371bf351b9#diff-68aa420a604c13765878c1fb39270f50ae0757b9b2f1b6609743632d2c7d0770R44.

that's ok then